File: python/fi/evals/manager_types.py
"""
Typed response models for :class:`~fi.evals.manager.EvalTemplateManager`.

Every model inherits from :class:`_APIResponse` which supports both
attribute access (``result.id``) and dict access (``result["id"]``)
for backwards compatibility.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict


# ---------------------------------------------------------------------------
# Dict-compatible base (Stripe / OpenAI SDK pattern)
# ---------------------------------------------------------------------------

class _APIResponse(BaseModel):
    """Base for all manager response models.

    Supports both attribute and dict-style access so existing code
    using ``result["id"]`` keeps working while new code gets IDE
    autocomplete via ``result.id``.
    """

    model_config = ConfigDict(extra="allow", populate_by_name=True)

    def __getitem__(self, key: str) -> Any:
        try:
            return getattr(self, key)
        except AttributeError:
            if self.model_extra and key in self.model_extra:
                return self.model_extra[key]
            raise KeyError(key)

    def get(self, key: str, default: Any = None) -> Any:
        try:
            return self[key]
        except KeyError:
            return default

    def __contains__(self, key: str) -> bool:
        return hasattr(self, key) or (
            bool(self.model_extra) and key in self.model_extra
        )


# ---------------------------------------------------------------------------
# Shared sub-models
# ---------------------------------------------------------------------------

class TemplateListItem(_APIResponse):
    id: str
    name: str
    template_type: Optional[str] = None
    eval_type: Optional[str] = None
    output_type: Optional[str] = None
    owner: Optional[str] = None
    created_by_name: Optional[str] = None
    version_count: Optional[int] = None
    current_version: Optional[str] = None
    last_updated: Optional[str] = None
    thirty_day_chart: Optional[List[Any]] = None
    thirty_day_error_rate: Optional[List[Any]] = None
    thirty_day_run_count: Optional[int] = None
    tags: Optional[List[str]] = None


class VersionItem(_APIResponse):
    id: str
    version_number: int
    is_default: bool = False
    criteria: Optional[str] = None
    model: Optional[str] = None
    config_snapshot: Optional[Dict[str, Any]] = None
    created_by_name: Optional[str] = None
    created_at: Optional[str] = None


class CompositeChildSummary(_APIResponse):
    child_id: str
    child_name: str
    order: int
    eval_type: Optional[str] = None
    pinned_version_id: Optional[str] = None
    pinned_version_number: Optional[int] = None
    weight: float = 1.0
    required_keys: Optional[List[str]] = None


class CompositeChildResult(_APIResponse):
    child_id: str
    child_name: str
    order: int
    score: Optional[Any] = None
    output: Optional[Any] = None
    reason: Optional[str] = None
    output_type: Optional[str] = None
    status: str = "completed"
    error: Optional[str] = None
    log_id: Optional[str] = None
    weight: float = 1.0
    error_localizer_result: Optional[Dict[str, Any]] = None


class GroundTruthItem(_APIResponse):
    id: str
    name: str
    description: Optional[str] = None
    file_name: Optional[str] = None
    columns: List[str] = []
    row_count: int = 0
    variable_mapping: Optional[Dict[str, str]] = None
    role_mapping: Optional[Dict[str, str]] = None
    embedding_status: str = "pending"
    embedded_row_count: Optional[int] = None
    storage_type: Optional[str] = None
    created_at: Optional[str] = None


class GroundTruthConfig(_APIResponse):
    enabled: bool = False
    ground_truth_id: Optional[str] = None
    mode: Optional[str] = None
    max_examples: Optional[int] = None
    similarity_threshold: Optional[float] = None
    injection_format: Optional[str] = None


# ---------------------------------------------------------------------------
# Template CRUD responses
# ---------------------------------------------------------------------------

class TemplateCreateResponse(_APIResponse):
    id: str
    name: str
    version: str


class TemplateDetailResponse(_APIResponse):
    id: str
    name: str
    description: Optional[str] = None
    template_type: Optional[str] = None
    eval_type: Optional[str] = None
    instructions: Optional[str] = None
    model: Optional[str] = None
    output_type: Optional[str] = None
    pass_threshold: Optional[float] = None
    choice_scores: Optional[Dict[str, float]] = None
    choices: Optional[Any] = None
    multi_choice: bool = False
    code: Optional[str] = None
    code_language: Optional[str] = None
    required_keys: Optional[List[str]] = None
    owner: Optional[str] = None
    created_by_name: Optional[str] = None
    version_count: Optional[int] = None
    current_version: Optional[str] = None
    tags: Optional[List[str]] = None
    check_internet: bool = False
    error_localizer_enabled: bool = False
    aggregation_enabled: Optional[bool] = None
    aggregation_function: Optional[str] = None
    composite_child_axis: Optional[str] = None
    config: Optional[Dict[str, Any]] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class TemplateUpdateResponse(_APIResponse):
    id: str
    name: Optional[str] = None
    updated: bool = True


class TemplateDuplicateResponse(_APIResponse):
    message: str
    eval_template_id: str


class TemplateListResponse(_APIResponse):
    items: List[TemplateListItem] = []
    total: int = 0
    page: int = 0
    page_size: int = 25


class BulkDeleteResponse(_APIResponse):
    deleted_count: int = 0


# ---------------------------------------------------------------------------
# Versioning responses
# ---------------------------------------------------------------------------

class VersionListResponse(_APIResponse):
    template_id: str
    versions: List[VersionItem] = []
    total: int = 0


class VersionCreateResponse(_APIResponse):
    id: str
    version_number: int
    is_default: bool = False


class VersionSetDefaultResponse(_APIResponse):
    id: str
    version_number: int
    is_default: bool = True


class VersionRestoreResponse(_APIResponse):
    id: str
    version_number: int
    is_default: bool = False
    restored_from: Optional[int] = None


# ---------------------------------------------------------------------------
# Composite responses
# ---------------------------------------------------------------------------

class CompositeCreateResponse(_APIResponse):
    id: str
    name: str
    template_type: str = "composite"
    aggregation_enabled: bool = True
    aggregation_function: Optional[str] = None
    composite_child_axis: Optional[str] = None
    children: List[CompositeChildSummary] = []


class CompositeDetailResponse(_APIResponse):
    id: str
    name: str
    description: Optional[str] = None
    template_type: str = "composite"
    aggregation_enabled: bool = True
    aggregation_function: Optional[str] = None
    composite_child_axis: Optional[str] = None
    children: List[CompositeChildSummary] = []
    tags: Optional[List[str]] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class CompositeExecutionResponse(_APIResponse):
    composite_id: str
    composite_name: Optional[str] = None
    aggregation_enabled: bool = True
    aggregation_function: Optional[str] = None
    aggregate_score: Optional[float] = None
    aggregate_pass: Optional[bool] = None
    children: List[CompositeChildResult] = []
    summary: Optional[Any] = None
    error_localizer_results: Optional[Dict[str, Any]] = None
    total_children: int = 0
    completed_children: int = 0
    failed_children: int = 0


# ---------------------------------------------------------------------------
# Ground Truth responses
# ---------------------------------------------------------------------------

class GroundTruthListResponse(_APIResponse):
    template_id: str
    items: List[GroundTruthItem] = []
    total: int = 0


class GroundTruthUploadResponse(_APIResponse):
    id: str
    name: str
    row_count: int = 0
    columns: List[str] = []
    embedding_status: str = "pending"


class GroundTruthDataResponse(_APIResponse):
    id: str
    page: int = 1
    page_size: int = 50
    total_rows: int = 0
    total_pages: int = 1
    columns: List[str] = []
    rows: List[Dict[str, Any]] = []


class GroundTruthStatusResponse(_APIResponse):
    id: str
    embedding_status: str = "pending"
    embedded_row_count: Optional[int] = None
    total_rows: int = 0
    progress_percent: Optional[float] = None


class GroundTruthVariableMappingResponse(_APIResponse):
    id: str
    variable_mapping: Dict[str, str] = {}


class GroundTruthRoleMappingResponse(_APIResponse):
    id: str
    role_mapping: Dict[str, str] = {}
    embedding_status: Optional[str] = None


class GroundTruthConfigResponse(_APIResponse):
    ground_truth: GroundTruthConfig


class GroundTruthSearchResponse(_APIResponse):
    query: str
    results: List[Dict[str, Any]] = []
    total: int = 0


class GroundTruthDeleteResponse(_APIResponse):
    deleted: bool = True
    id: str


# ---------------------------------------------------------------------------
# Usage / Feedback / Charts responses
# ---------------------------------------------------------------------------

class TemplateUsageResponse(_APIResponse):
    template_id: str
    stats: Dict[str, Any] = {}
    chart: List[Any] = []
    logs: Optional[Any] = None  # paginated dict {items, page, page_size}


class TemplateFeedbackListResponse(_APIResponse):
    template_id: str
    items: List[Dict[str, Any]] = []
    total: int = 0
    page: int = 0
    page_size: int = 25


class TemplateChartsResponse(_APIResponse):
    charts: Dict[str, Any] = {}


# ---------------------------------------------------------------------------
# Playground response
# ---------------------------------------------------------------------------

class PlaygroundRunResponse(_APIResponse):
    output: Optional[Any] = None
    reason: Optional[str] = None
    model: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None
    output_type: Optional[str] = None
    log_id: Optional[str] = None

File: typescript/agent-learning-kit/src/manager-types.ts
/**
 * Typed response interfaces for {@link EvalTemplateManager}.
 *
 * All interfaces use snake_case field names to match the backend JSON
 * wire format. Both `result.id` and `result["id"]` work natively in
 * TypeScript (no special base class needed).
 */

// ---------------------------------------------------------------------------
// Shared sub-types
// ---------------------------------------------------------------------------

export interface TemplateListItem {
    id: string;
    name: string;
    template_type?: string;
    eval_type?: string;
    output_type?: string;
    owner?: string;
    created_by_name?: string;
    version_count?: number;
    current_version?: string;
    last_updated?: string;
    thirty_day_chart?: any[];
    thirty_day_error_rate?: any[];
    thirty_day_run_count?: number;
    tags?: string[];
}

export interface VersionItem {
    id: string;
    version_number: number;
    is_default: boolean;
    criteria?: string;
    model?: string;
    config_snapshot?: Record<string, any>;
    created_by_name?: string;
    created_at?: string;
}

export interface CompositeChildSummary {
    child_id: string;
    child_name: string;
    order: number;
    eval_type?: string;
    pinned_version_id?: string | null;
    pinned_version_number?: number | null;
    weight: number;
    required_keys?: string[];
}

export interface CompositeChildResult {
    child_id: string;
    child_name: string;
    order: number;
    score?: any;
    output?: any;
    reason?: string;
    output_type?: string;
    status: string;
    error?: string | null;
    log_id?: string | null;
    weight: number;
    error_localizer_result?: Record<string, any> | null;
}

export interface GroundTruthItemResponse {
    id: string;
    name: string;
    description?: string;
    file_name?: string;
    columns: string[];
    row_count: number;
    variable_mapping?: Record<string, string> | null;
    role_mapping?: Record<string, string> | null;
    embedding_status: string;
    embedded_row_count?: number;
    storage_type?: string;
    created_at?: string;
}

export interface GroundTruthConfigData {
    enabled: boolean;
    ground_truth_id?: string | null;
    mode?: string;
    max_examples?: number;
    similarity_threshold?: number;
    injection_format?: string;
}

// ---------------------------------------------------------------------------
// Template CRUD
// ---------------------------------------------------------------------------

export interface TemplateCreateResponse {
    id: string;
    name: string;
    version: string;
}

export interface TemplateDetailResponse {
    id: string;
    name: string;
    description?: string;
    template_type?: string;
    eval_type?: string;
    instructions?: string;
    model?: string;
    output_type?: string;
    pass_threshold?: number;
    choice_scores?: Record<string, number> | null;
    choices?: any;
    multi_choice?: boolean;
    code?: string | null;
    code_language?: string | null;
    required_keys?: string[];
    owner?: string;
    created_by_name?: string;
    version_count?: number;
    current_version?: string;
    tags?: string[];
    check_internet?: boolean;
    error_localizer_enabled?: boolean;
    aggregation_enabled?: boolean;
    aggregation_function?: string;
    composite_child_axis?: string;
    config?: Record<string, any> | null;
    created_at?: string;
    updated_at?: string;
}

export interface TemplateUpdateResponse {
    id: string;
    name?: string;
    updated: boolean;
}

export interface TemplateDuplicateResponse {
    message: string;
    eval_template_id: string;
}

export interface TemplateListResponse {
    items: TemplateListItem[];
    total: number;
    page: number;
    page_size: number;
}

export interface BulkDeleteResponse {
    deleted_count: number;
}

// ---------------------------------------------------------------------------
// Versioning
// ---------------------------------------------------------------------------

export interface VersionListResponse {
    template_id: string;
    versions: VersionItem[];
    total: number;
}

export interface VersionCreateResponse {
    id: string;
    version_number: number;
    is_default: boolean;
}

export interface VersionSetDefaultResponse {
    id: string;
    version_number: number;
    is_default: boolean;
}

export interface VersionRestoreResponse {
    id: string;
    version_number: number;
    is_default: boolean;
    restored_from?: number;
}

// ---------------------------------------------------------------------------
// Composite
// ---------------------------------------------------------------------------

export interface CompositeCreateResponse {
    id: string;
    name: string;
    template_type: string;
    aggregation_enabled: boolean;
    aggregation_function?: string;
    composite_child_axis?: string;
    children: CompositeChildSummary[];
}

export interface CompositeDetailResponse extends CompositeCreateResponse {
    description?: string;
    tags?: string[];
    created_at?: string;
    updated_at?: string;
}

export interface CompositeExecutionResponse {
    composite_id: string;
    composite_name?: string;
    aggregation_enabled: boolean;
    aggregation_function?: string;
    aggregate_score?: number | null;
    aggregate_pass?: boolean | null;
    children: CompositeChildResult[];
    summary?: any;
    error_localizer_results?: Record<string, any> | null;
    total_children: number;
    completed_children: number;
    failed_children: number;
}

// ---------------------------------------------------------------------------
// Ground Truth
// ---------------------------------------------------------------------------

export interface GroundTruthListResponse {
    template_id: string;
    items: GroundTruthItemResponse[];
    total: number;
}

export interface GroundTruthUploadResponse {
    id: string;
    name: string;
    row_count: number;
    columns: string[];
    embedding_status: string;
}

export interface GroundTruthDataResponse {
    id: string;
    page: number;
    page_size: number;
    total_rows: number;
    total_pages: number;
    columns: string[];
    rows: Array<Record<string, any>>;
}

export interface GroundTruthStatusResponse {
    id: string;
    embedding_status: string;
    embedded_row_count?: number;
    total_rows: number;
    progress_percent?: number;
}

export interface GroundTruthVariableMappingResponse {
    id: string;
    variable_mapping: Record<string, string>;
}

export interface GroundTruthRoleMappingResponse {
    id: string;
    role_mapping: Record<string, string>;
    embedding_status?: string;
}

export interface GroundTruthConfigResponse {
    ground_truth: GroundTruthConfigData;
}

export interface GroundTruthSearchResponse {
    query: string;
    results: Array<Record<string, any>>;
    total: number;
}

export interface GroundTruthDeleteResponse {
    deleted: boolean;
    id: string;
}

// ---------------------------------------------------------------------------
// Usage / Feedback / Charts
// ---------------------------------------------------------------------------

export interface TemplateUsageResponse {
    template_id: string;
    stats: Record<string, any>;
    chart: any[];
    logs?: any;
}

export interface TemplateFeedbackListResponse {
    template_id: string;
    items: Array<Record<string, any>>;
    total: number;
    page: number;
    page_size: number;
}

export interface TemplateChartsResponse {
    charts: Record<string, any>;
}

// ---------------------------------------------------------------------------
// Playground
// ---------------------------------------------------------------------------

export interface PlaygroundRunResponse {
    output?: any;
    reason?: string;
    model?: string | null;
    metadata?: Record<string, any> | null;
    output_type?: string;
    log_id?: string;
}
